### ISSUE TITLE
Feature Request : Search Page, HTML Head Items and SEO Sitemap

### ISSUE THREAD

**Issue by ramesh-dada** [robiso/wondercms#237]
I want to add some codes in head tag in order to improve seo I mean I have to add schema tags and all that, I also need Search Page in order to search whole website for the query, and at last I need Sitemap.xml as It is very important for indexing.

**Comment by robiso** [comment]
@ramesh-dada
1. Adding some better HTML head items is as easy as including this in your theme:
```
	<head>
        <!-- Encoding, browser compatibility, viewport -->
    		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    		<meta http-equiv="X-UA-Compatible" content="IE=edge">
    		<meta name="viewport" content="width=device-width, initial-scale=1">

        <!-- Search Engine Optimization (SEO) -->
    		<meta name="title" content="<?= $Wcms->get('config', 'siteTitle') ?> - <?= $Wcms->page('title') ?>" />
    		<meta name="description" content="<?= $Wcms->page('description') ?>">
    		<meta name="keywords" content="<?= $Wcms->page('keywords') ?>">
    		<meta property="og:url" content="<?= $this->url() ?>" />
    		<meta property="og:type" content="website" />
    		<meta property="og:site_name" content="<?= $Wcms->get('config', 'siteTitle') ?>" />
    		<meta property="og:title" content="<?= $Wcms->page('title') ?>" />
    		<meta name="twitter:site" content="<?= $this->url() ?>" />
    		<meta name="twitter:title" content="<?= $Wcms->get('config', 'siteTitle') ?> - <?= $Wcms->page('title') ?>" />
    		<meta name="twitter:description" content="<?= $Wcms->page('description') ?>" />
    		<title><?= $Wcms->get('config', 'siteTitle') ?> - <?= $Wcms->page('title') ?></title>

        <!-- Theme CSS -->
    		<link rel="preload stylesheet" as="style" href="<?= $Wcms->asset('css/style.css') ?>">
    		
        <!-- Admin CSS -->
    		<?= $Wcms->css()?>

	</head>
```

2. Sitemap plugin is available here: https://github.com/StephanStanisic/simple-seo
3. Searching is a specific challenge, if anyone would be interested in taking a look for developing this, I'd love to give it a try.

**Comment by robiso** [comment]
@DaddyMadu I've updated the Simple SEO plugin and forked it from the original author Stephan Stanisic.

It's now included in the official list of plugins, so make sure to uninstall the one you have and remove it completely from your website and then installing it from the Settings -> Plugins.

We try not to include too much functionality into WonderCMS to retain it's mission for small size, thus we use plugins and depend on maintainers/contributors, as the "core" should be as clean as possible without having to maintain too much different aspects.

**Comment by DaddyMadu** [comment]
@robiso thank you very much for that swifty replay! 
i know how what you are talking about but such functionality nowadays is a must in any basic installation. 
any way thanks for taking care of it, and hope the best for you guys out there
you have such an amazing project keep it up!

**Comment by robiso** [comment]
You're welcome @DaddyMadu, but please note that everyone has an opinion on what are the "basic" things that should be included in WonderCMS.

We try to be impartial and try to provide/support the community with plugins and let the users decide on what they would want included.
The other benefit here is maintenance, as the "core team" is trying to keep the WonderCMS core up to date and safe, without being too distracted with themes and plugins. 

Appreciate the support and thanks for your feedback.

**Comment by joho1968** [comment]
As for search ... I think it'd be a good idea to implement some hooks for this.

- It would be theme independent
- It won't break anything for themes that don't support it
- It allows themes to place a "search box" wherever

Something like "search_paint" or "paint_search" maybe, if you want to have other "paint_"-like hooks for other things in the future.

I can certainly give the search thing a try, and try to implement it in my [Bootstrap 5 theme for WonderCMS](https://github.com/joho1968/wcms-bs5).

**Comment by joho1968** [comment]
Come to think of it, it probably needs to be the other way around. The theme should implement this, and then be able to call a public core function (?) that returns an array that is either empty, or matching content. The array could contain pairs, like ['thisistheurl' => 'matching text highlight'], or pairs like ['url' => 'page title'], or just a simple a list of URLs. But the page title would be nice to get for display.

**Comment by joho1968** [comment]
The latest version of the WonderCMS BS5 theme now includes a search function that will search WonderCMS content as well as the Simple Blog plugin content.

**Comment by robiso** [comment]
@joho1968 this sounds pretty AMAZING. Will check it out if we can somehow include it either natively or via a plugin. Amazing effort :)

**Comment by joho1968** [comment]
Yeah, I think native search hooks would be a great idea. I do think it should be in core. I mean, if you look at my code in the BS5 theme, it's really not that many lines of code, nor is it very complicated :blush: I'll keep it in my theme until you make a decision on core/native functionality.

**Comment by robiso** [comment]
@joho1968 @DaddyMadu @ramesh-dada, this is now in a pull request.

More details here:
https://github.com/WonderCMS/wondercms/pull/327
